core/src/test/scala/unit/kafka/network/SocketServerTest.scala

```diff
@@ -593,4 +596,278 @@ class SocketServerTest extends JUnitSuite {
+    selector.failures.clear()
+    val socket = connect(testableServer)
+    val serializedBytes = producerRequestBytes
+    sendRequest(socket, serializedBytes)
+    val requestChannel = testableServer.requestChannel
+
+    val request = receiveRequest(requestChannel)
+    val connectionId = request.context.connectionId
+    processRequest(requestChannel, request)
+    assertEquals(serializedBytes.toSeq, receiveResponse(socket).toSeq)
+    assertNotNull("Channel should not have been closed", selector.channel(connectionId))
+    assertNull("Channel should not be closing", selector.closingChannel(connectionId))
+  }
+
+  class TestableSocketServer extends SocketServer(KafkaConfig.fromProps(props), new Metrics, Time.SYSTEM, credentialProvider) {
+
+    @volatile var selector: Option[TestableSelector] = None
+
+    override def newProcessor(id: Int, connectionQuotas: ConnectionQuotas, listenerName: ListenerName,
+                                protocol: SecurityProtocol, memoryPool: MemoryPool): Processor = {
+      new Processor(id, time, config.socketRequestMaxBytes, requestChannel, connectionQuotas,
+        config.connectionsMaxIdleMs, listenerName, protocol, config, metrics, credentialProvider, memoryPool) {
+        override protected[network] def createSelector(maxReceiveSize: Int, connectionMaxIdleMs: Long,
+             metrics: Metrics, time: Time, metricGrpPrefix: String, metricTags: java.util.Map[String, String],
+             metricsPerConnection: Boolean, recordTimePerConnection: Boolean,
+             channelBuilder: ChannelBuilder, memoryPool: MemoryPool): Selector = {
+
+           val testableSelector = new TestableSelector(maxReceiveSize, connectionMaxIdleMs, metrics, time, metricGrpPrefix, metricTags,
+                            metricsPerConnection, recordTimePerConnection, channelBuilder, memoryPool)
+           selector = Some(testableSelector)
+           testableSelector
+        }
+      }
+    }
+    def testableSelector = selector.getOrElse(throw new IllegalStateException("Selector not created"))
+
+    def waitForChannelException(connectionId: String, socket: Socket, operation: SelectorOperation.Value): Unit = {
+      val selector = testableSelector
+      selector.waitForOperations(operation, 1)
+      TestUtils.waitUntilTrue(() =>  selector.allLocallyClosedChannels.contains(connectionId), "Channel not closed")
+      assertEquals(0, connectionCount(socket.getLocalAddress))
+      assertTrue("Inflight responses not cleared", processor(0).inflightResponses.isEmpty)
+      assertNull("Channel not removed", selector.channel(connectionId))
+      assertNull("Closing channel not removed", selector.closingChannel(connectionId))
+      assertTrue("Unexpected disconnect notification",testableSelector.allDisconnectedChannels.isEmpty)
+    }
+  }
+
+  object SelectorOperation extends Enumeration {
+    type SelectorOperation = Value
+    val REGISTER, POLL, SEND, MUTE, UNMUTE, CLOSE, DISCONNECTED = Value
```

Comment:
Not sure if it matters here, but Scala enumerations are usually discouraged because they don't have exhaustivity warnings. Sealed trait + case objects are the usual approach.

Reply:
Switched to sealed trait.